### ISSUE TITLE
Create framework verify config distribution has taken place.

### ISSUE THREAD

**Issue by mandarjog** [istio/istio#14629]
May e2e test follow the following pattern

1. apply some configuration
2. wait()
3. run test and verify behaviour

The wait leads to flakiness of we wait too little before starting test, and becomes too slow if we wait for too long.

We would like a mechanism to ensure that a config applied to api server has reached an envoy.

There are several options

1. Break applying config in 2 phases. 
    a. phase 1 send the config that is required by test
    b. phase 2 write a sentinel config. A service entry that uses a guuid as a host name is a good sentinel.

    We can test the proxies in questions to ensure that the service entry has reached them by looking for the sentinel.
    We can do this is a tigher loop so that the successful polls are quicker.
2. Use metrics from pilot and envoy

If sentinel has not reached proxy by the timeout, check if the sentinel has reached pilot by asking for a config dump

**Comment by mandarjog** [comment]
@kyessenov @douglas-reid

**Comment by kyessenov** [comment]
1.b is a hack and is going to cause more problems than fixes things I think. It doesn't work for non-networking config (policy/telemtry/CA).

**Comment by douglas-reid** [comment]
@kyessenov non-networking config is (perhaps) easier to validate as being active. for instance, we can retrieve the counts of the number of `handlers`, `instances`, and `rules` in use by mixers -- and then poll until we see that change to reflect new config (mixer config tends to be 1:1 with operator provided config). That gives us a way of knowing when policy/telemetry rules should be in effect. I think some of the e2e tests do this already, in some form.

the difficulty, as i understand it, is for the networking config at the proxy level. we don't have a good way of asking the proxies if they received the new config. a sentinel gives us some approximation, but is certainly not sufficient in isolation.

**Comment by kyessenov** [comment]
@douglas-reid I agree that it _would_ work but it is a hack. It's using features for their non-intended use, leading to unexpected behaviors and complexity. We already have tons of complexity due to the underlying issue, so a better solution would be a way to simplify things than making them more complicated.

**Comment by esnible** [comment]
I have been thinking about a similar problem.  Pilot sometimes generates configuration that is rejected by an Envoy.  It is hard for a user to detect this (we have a Grafana dashboard an an istioctl subcommand).  Once it happens further rule changes elsewhere don't take effect.

I propose Istio rules have a Status field.  When a rule arrives it is tagged *provisional*.  If all of the Envoys accept XDS after a change, the provisional flags are a removed and changed to *accepted*.  After XDS push is rejected by Envoy any *provisional* rules are changed to *dubious*.  (When new K8s config arrives we could try to generate XDS with and without the dubious and offer them to Envoy; hoping eventually that everything is accepted).

The same provisional/accepted could be used to speed up the tests.

**Comment by howardjohn** [comment]
This is a duplicate of https://github.com/istio/istio/issues/6170, lets track in a single place
